### PR TITLE
fix: Update volume material mapping

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
This PR update the volume material mapping to solve 2 issue with it.

-First it had the same flaw as the one from the surface mapping (#1088)
-Second all the material hits where kept until the end of the mapping process, this meant that if the detector was too big and too many material track were used it could run out of memory. The are now progressively added to the grid during the mapping and averaged at the end.

Those change shouldn't affect the result mapping itself. (it was tested again with the ATLAS detector and appear to be fine)

### CODE REVIEW

**Comment by paulgessinger** [comment]
Could we add a reproducibility test at the python level with the ODD?

**Comment by Corentin-Allaire** [comment]
> Could we add a reproducibility test at the python level with the ODD?

The issue is that the ODD doesn't really have volume that work well with volume mapping. I could try to think about it though...

**Comment by paulgessinger** [comment]
I guess for a reproducibility test it wouldn't really matter. Just have the volume mapping map the entire detector, and make sure it does the same thing as it did before. 

Or do you think that doesn't tell us anything?

**Comment by Corentin-Allaire** [comment]
> I guess for a reproducibility test it wouldn't really matter. Just have the volume mapping map the entire detector, and make sure it does the same thing as it did before.
> 
> Or do you think that doesn't tell us anything?

Sure I can do that, I am not sure we would catch change there that we would with the surface mapping (except direct change to the volume mapping) but it doesn't cost much to add one

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 1]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+    std::pair<const GeometryIdentifier, BinUtility>& currentBinning,
+    Acts::MaterialSlab properties, Vector3 position, Vector3 direction) const {
+  if (currentBinning.second.dimensions() == 0) {
+    // Writing homogeneous material for the current volumes no need to create
+    // extra hits. We directly accumulate the material
+    mState.homogeneousGrid[currentBinning.first].accumulate(properties);
+    return;
+  }
 
+  // Computing the extra hits properties based on the mappingStep lenght
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:247]:
```suggestion
  // Computing the extra hits properties based on the mappingStep length
```

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 2]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+      } else {
+        throw std::domain_error("No grind 3D was found");
+      }
+    }
   }
-  matPoint.push_back(std::pair(properties, extraPosition));
 
   if (remainder > 0) {
-    // adjust the thickness of the last extrapolated step
+    // We need to had an additional extra hit with the remainder lengh. Adjust
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:281]:
```suggestion
    // We need to had an additional extra hit with the remainder length. Adjust
```

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 3]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+
+    if (currentBinning.second.dimensions() == 2) {
+      auto grid = mState.grid2D.find(currentBinning.first);
+      if (grid != mState.grid2D.end()) {
+        // Find which grid bin the material fall into then accumulate
+        Acts::Grid2D::index_t index = grid->second.localBinsFromLowerLeftEdge(
+            mState.transform2D[currentBinning.first](extraPosition));
+        grid->second.atLocalBins(index).accumulate(properties);
+      } else {
+        throw std::domain_error("No grind 2D was found");
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:265]:
```suggestion
        throw std::domain_error("No grid 2D was found");
```

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 4]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+      }
+    } else if (currentBinning.second.dimensions() == 3) {
+      auto grid = mState.grid3D.find(currentBinning.first);
+      if (grid != mState.grid3D.end()) {
+        // Find which grid bin the material fall into then accumulate
+        Acts::Grid3D::index_t index = grid->second.localBinsFromLowerLeftEdge(
+            mState.transform3D[currentBinning.first](extraPosition));
+        grid->second.atLocalBins(index).accumulate(properties);
+      } else {
+        throw std::domain_error("No grind 3D was found");
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:275]:
```suggestion
        throw std::domain_error("No grid 3D was found");
```

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 5]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+    Vector3 extraPosition = position + volumeStep * direction;
+    if (currentBinning.second.dimensions() == 2) {
+      auto grid = mState.grid2D.find(currentBinning.first);
+      if (grid != mState.grid2D.end()) {
+        // Find which grid bin the material fall into then accumulate
+        Acts::Grid2D::index_t index = grid->second.localBinsFromLowerLeftEdge(
+            mState.transform2D[currentBinning.first](extraPosition));
+        grid->second.atLocalBins(index).accumulate(properties);
+      } else {
+        throw std::domain_error("No grind 2D was found");
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:293]:
```suggestion
        throw std::domain_error("No grid 2D was found");
```

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 6]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+      }
+    } else if (currentBinning.second.dimensions() == 3) {
+      auto grid = mState.grid3D.find(currentBinning.first);
+      if (grid != mState.grid3D.end()) {
+        // Find which grid bin the material fall into then accumulate
+        Acts::Grid3D::index_t index = grid->second.localBinsFromLowerLeftEdge(
+            mState.transform3D[currentBinning.first](extraPosition));
+        grid->second.atLocalBins(index).accumulate(properties);
+      } else {
+        throw std::domain_error("No grind 3D was found");
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:303]:
```suggestion
        throw std::domain_error("No grid 3D was found");
```

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 7]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+      auto grid = mState.grid2D.find(matBin.first);
+      if (grid != mState.grid2D.end()) {
+        Acts::MaterialGrid2D matGrid = mapMaterialPoints(grid->second);
+        MaterialMapper<Acts::MaterialGrid2D> matMap(
+            mState.transform2D[matBin.first], matGrid);
+        mState.volumeMaterial[matBin.first] = std::make_unique<
+            InterpolatedMaterialMap<MaterialMapper<Acts::MaterialGrid2D>>>(
+            std::move(matMap), matBin.second);
+      } else {
+        throw std::domain_error("No grind 2D was found");
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:332]:
```suggestion
        throw std::domain_error("No grid 2D was found");
```

File: Core/src/Material/VolumeMaterialMapper.cpp [hunk 8]
```diff
@@ -206,68 +234,118 @@ void Acts::VolumeMaterialMapper::collectMaterialSurfaces(
...
+      auto grid = mState.grid3D.find(matBin.first);
+      if (grid != mState.grid3D.end()) {
+        Acts::MaterialGrid3D matGrid = mapMaterialPoints(grid->second);
+        MaterialMapper<Acts::MaterialGrid3D> matMap(
+            mState.transform3D[matBin.first], matGrid);
+        mState.volumeMaterial[matBin.first] = std::make_unique<
+            InterpolatedMaterialMap<MaterialMapper<Acts::MaterialGrid3D>>>(
+            std::move(matMap), matBin.second);
+      } else {
+        throw std::domain_error("No grind 3D was found");
```
Comment (paulgessinger) [inline Core/src/Material/VolumeMaterialMapper.cpp:347]:
```suggestion
        throw std::domain_error("No grid 3D was found");
```

**Review by paulgessinger — COMMENTED** [review]
Some typo fixes. Other than that this looks good as far as I can tell.

**Comment by Corentin-Allaire** [comment]
> Some typo fixes. Other than that this looks good as far as I can tell.

Thanks for finding the typos @paulgessinger

**Review by paulgessinger — APPROVED** [review]

**Pull request merged**